### PR TITLE
RC_Channel: add hysteresis in range_to_pwm to remove input jitter

### LINKED ISSUE #2716: Plane: rcin throttle jitter results in large throttle bumps
That happens in FBWB mode, should be the same in CRUISE and AUTO with throttle nudge.
Input jitter is 1us. Output jitter is quite significant ~70us peak to peak which is almost 10% of TECS' throttle range.

After some digging into the problem I found two causes.
1. 1us jitter may result 1% jitter because the throttle is converted to 0-100 integer range.
2. TECS  reacts quite aggressively to a minor speed change fed to update_pitch_throttle.

Configuration info:
ArduPlane 3.3.0, PX4, s.bus connection to RX,
airspeed sensor enabled

### PR DESCRIPTION
**Pull request by KIrill-ka**
This fixes #2716.
The idea is to use different threshold values whether input increases or decreases.
For example, RC input range is 1000-2000, desired range is 0-100. When input increases from 1499 to 1500, output will change from 49 to 50%, but  it won't get back to 49% until input goes down to 1495.

### CODE REVIEW

**Comment by tridge** [comment]
I think I'd rather just change the calculation to use a float. Just making control_in be a float should fix this issue

**Comment by KIrill-ka** [comment]
I tried to increase throttle resolution and it didn't help. TECS seems to react aggressively even if throttle change is very small (0.1%). I used 0-1000 integer range, not float.

**Comment by KIrill-ka** [comment]
Below is an illustration how it works with high resolution throttle input in FBWB mode.
Interesting thing is that there's no output jitter in FBWA mode unless input jitter crosses % boundary.
aa

**Comment by magicrub** [comment]
I agree, changing this to float will be nice. It will come in handy when Plane reverse thrust comes in later.

File: libraries/RC_Channel/RC_Channel.cpp [hunk 1]
```diff
@@ -337,12 +338,18 @@ RC_Channel::pwm_to_range_dz(uint16_t dead_zone)
...
-        return _low;
+    if (r_in > radio_trim_low) {
+        int32_t v;
+        int16_t d;
+
+        d = radio_max - radio_trim_low;
+        v = (int32_t)(_high - _low) * (int32_t)(r_in - radio_trim_low) + d/2;
+        ret = _low + v / d;
+        if (ret - _prev_control_in == 1 && v % d < d/2) ret--;
+    } else if (dead_zone > 0) ret = 0;
```
Comment (magicrub) [inline libraries/RC_Channel/RC_Channel.cpp]: the else if ret = 0 needs  { }

File: libraries/RC_Channel/RC_Channel.cpp [hunk 2]
```diff
@@ -337,12 +338,18 @@ RC_Channel::pwm_to_range_dz(uint16_t dead_zone)
...
-    else
-        return _low;
+    if (r_in > radio_trim_low) {
+        int32_t v;
+        int16_t d;
+
+        d = radio_max - radio_trim_low;
+        v = (int32_t)(_high - _low) * (int32_t)(r_in - radio_trim_low) + d/2;
+        ret = _low + v / d;
+        if (ret - _prev_control_in == 1 && v % d < d/2) ret--;
```
Comment (magicrub) [inline libraries/RC_Channel/RC_Channel.cpp]: ret-- needs {}

File: libraries/RC_Channel/RC_Channel.cpp [hunk 3]
```diff
@@ -337,12 +338,18 @@ RC_Channel::pwm_to_range_dz(uint16_t dead_zone)
...
+    } else if (dead_zone > 0) ret = 0;
+    else ret = _low;
```
Comment (magicrub) [inline libraries/RC_Channel/RC_Channel.cpp]: ret = _low needs { }

**Comment by KIrill-ka** [comment]
Tom, are you going to integrate the patch? If yes, I'll fix code styling.

**Comment by magicrub** [comment]
Sorry for the long delay in looking at this PR. Tridge and Randy are the gate keepers, I'm just code reviewing to help them out - I can't merge it myself. We're rolling out a new version of plane soon so I'm helping out by going through the PRs to make it easier on them, getting some extra sure balls on it. The more perfect the code, the easier it is to accept it and merge it in. If they have to make edits to conform to coding standards or to improve readability then the chances go up that it gets back burnerd.

Speaking of readability, can you put 
(ret - _prev_control_in == 1 && v % d < d/2
On separate lines? That's a hard line to read.

The comments on the brackets are for the style
http://dev.ardupilot.com/wiki/style-guide/

Thanks!

**Comment by SlipstreamFPV** [comment]
Was this implemented? I'm suffering from output jitter in cruise mode as well. I guess its a simple test to disable throttle nudge and see if the throttle noise is causing it.

**Comment by magicrub** [comment]
This was not merged. It was requested that it be changed to use floats which would greatly simply it both mathematically and readability.

**Comment by KIrill-ka** [comment]
> It was requested that it be changed to use floats which would greatly simply it both 
> mathematically and readability.

Note that floats alone won't solve the problem. RC input is always discrete. 
> I tried to increase throttle resolution and it didn't help.

**Comment by SlipstreamFPV** [comment]
This issue is still plaguing my plane. On some throttle percentages my plane is fine, on others the pitch servo and throttle jump up and down because my taranis has a bit of noise on the well used gimbals. In manual it's not an issue because it's only 5-10us which doesn't make much change but in auto modes the pixhawk is seeing a requested change and trying to execute it which leads to throttle pulses of multiple percentage points up and down. Surely there's an easy way to add dampening to this in auto modes. I can't imagine anyone actually changes there throttle by 10us while flying auto modes and expects a change. What about a simple 1 second average reading while in auto modes?

**Comment by magicrub** [comment]
There's throttle slew to address a :1 second average"  but won't help on
tiny changes that you're experiencing.

**Comment by SlipstreamFPV** [comment]
I didn't explain that last bit very well at all! What about an adjustable averaging on the throttle input channel while in auto modes to prevent the autopilot seeing the tiny fluctuations? Or a minimum change in throttle channel us before the autopilot will respond in auto modes?

**Comment by amilcarlucas** [comment]
@SlipstreamFPV can you rebase this?

**Comment by WickedShell** [comment]
Closing this for now as it seems to be pretty thoroughly abandoned, the proposed solution would introduce control problems on planes that need small movement on channels (IE this would break a lot of manual deepstall planes pitch input). The core problem to be resolved here is that internally we map throttle onto an integer percentage range, which does need to be fixed, but is not what this PR is actually attempting to do. There may also be a different requirement to fix TECs reaction to step input changes, but again that's not what this is attempting to do.

**Comment by KIrill-ka** [comment]
Well, no problem. This is just an odd contribution I have no personal benefit of. And I kinda understand why it's not merged in. However, I can't resist to comment on this, because I don't see a single real argument (see below). 

> Closing this for now as it seems to be pretty thoroughly abandoned

I'm still here and the changes were rebased upon request.

> , the proposed solution would introduce control problems on planes that need small movement on channels (IE this would break a lot of manual deepstall planes pitch input).

I don't see how. Can you prove it?

> The core problem to be resolved here is that internally we map throttle onto an integer percentage range, which does need to be fixed,

It was long time ago I was experimenting on this... Increasing throttle resolution didn't seem to cure the problem. The source throttle signal is integer and converting it to a floating point won't remove the jitter.

> There may also be a different requirement to fix TECs reaction to step input changes, but again that's not what this is attempting to do.

Yes, but the proposed change is to remove the jitter noise introduced by the RC system. It's not something input by the operator. It's not clear why this should be a function of TECS.

**Pull request closed without merging**